### PR TITLE
Play silence to keep audio device awake

### LINKED ISSUE #14386: Play silence to prevent missing information at start of speech
### Is your feature request related to a problem? Please describe.
When using some bluetooth headphones, speech output is not heard after NVDA has some moments of silence.
The headphones appear to go into standby and miss the start of speech.
Keeping NVDA active is tiring, missing speech is annoying and confusing.

### Describe the solution you'd like
Add an option to play silence to keep NVDA active.

### Describe alternatives you've considered
Use [bluetooth audio add-on](https://addons.nvda-project.org/addons/BluetoothAudio.en.html).

### Additional context
- This is a problem that is likely to be common to many users, bluetooth headsets / airpods are very common.
- This feature is available in Jaws since 2020.

### PR DESCRIPTION
**Pull request by mltony**
### Link to issue number:
Fixes #14386
### Summary of the issue:
Play silence in order to keep audio device open
### Description of user facing changes
By default most users won't notice anything as default volume is 0 (playing silence).
I added two config options in the advanced panel: silence duration (default = 30 seconds) and white noise volume (default is set to 0). Playing white noise would be helpfulin order to debug any user issues related to audio dropout.
### Description of development approach
Most of heavylifting has  been done by @jcsteh.
In `wasapi.cpp` there has been added `class SilencePlayer` that spawns a new thread, that just plays silence/white noise when requested. Requests are coming from python code in `nvwave.py` from `WasapiWavePlayer.feed()` function.

### Testing strategy:
Tested with white noise that it is being played as expected. Tested that config changes reflect immediately.
Couldn't test with silence as apparently my bluetooth headphones are not suffering from dropout problem in the first place.
### Known issues with pull request:
N/A
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by cary-rowen** [comment]
I'll test this after you fix the CI build fail

File: nvdaHelper/local/wasapi.cpp [hunk 1]
```diff
@@ -187,7 +190,6 @@ class WasapiPlayer {
 	HRESULT pause();
 	HRESULT resume();
 	HRESULT setChannelVolume(unsigned int channel, float level);
-
```
Comment (lukaszgo1) [inline nvdaHelper/local/wasapi.cpp:190]: Why this change?

File: user_docs/en/userGuide.t2t [hunk 2]
```diff
@@ -2594,6 +2594,16 @@ This option allows you to specify if NVDA will play an error sound in case an er
...
 
+==== Silence duration ====[silenceDuration]
+
+This edit box specifies how long NVDA plays silence after every utterance. We play silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
+
+You can specify silence duration to be 0 in order to disable this feature.
+
+==== White noise volume slider ====[whiteNoiseVolume]
+
+When you set white noise volume to any value above zero, then instead of silence NVDA would playh white noise in order to keep audio device open. This feature is useful for debugging audio issues.
```
Comment (lukaszgo1) [inline user_docs/en/userGuide.t2t:2605]:
```suggestion
When you set white noise volume to any value above zero, then instead of silence NVDA would play white noise in order to keep audio device open. This feature is useful for debugging audio issues.
```

File: source/core.py
Comment (lukaszgo1) [inline source/core.py:1]: Please re-initialize `nvwave` when user restores config to saved / defaults.

**Comment by CyrilleB79** [comment]
To me, these options seem too specific / technical for most of the users, including some users impacted by bluetooth device audio cut-off. Not every user knows what a white noise is, and from a user point of view, playing silence or with noise is not very meaningful.

> Playing white noise would be helpfulin order to debug any user issues related to audio dropout.

First of all, if white noise is only for debug purpose, it should not be in audio settings IMO, maybe just in advanced settings, or better, in an add-on.
If instead white noise is required by some device or sound cards, I think that we should just enforce white noise for everybody. In any case this white noise should not be audible, so it should not make a difference with full silence.

Personally, I would see just one option "Duration of audio inactivity (in seconds) before closing the audio stream". Or even simpler: a checkbox "Keep the audio device/stream always open". Then the in the User Guide, give information to the user to help setting this value, i.e.:
* too low = audio may be cut-off more often with a Bluetooth device
* too high = the battery discharges faster.

File: source/core.py
Comment (mltony) [reply source/core.py:1]: No need to reinitialize since it reloads current settings for duration and volume automatically.

**Comment by mltony** [comment]
@cary-rowen,
Feel free to test this now. installer NVDA test ids failing for everyone.

@CyrilleB79 

> these options seem too specific / technical for most of the users
> it should not be in audio settings 

I already put both settings in advanced panel. Most users should be happy with default values.

Some users complained to Jamie that this patch didn't fix their audio issue. We can either try to guess what happened to those users. Or alternatively we can ask them to switch to white noise and this will give us more information, because:
1. If white noise is not audible - something is really broken.
2. If white noise is audible but from another audio output device, then there is a problem with synchronization.
3. If white noise is audible from correct audio output, then something is wrong with auido 
If you take away white noise option, then there won't be a good way to debug those issues. 

> • too low = audio may be cut-off more often with a Bluetooth device
> • too high = the battery discharges faster.

Updated doc.

**Comment by CyrilleB79** [comment]
Sorry, @mltony, I had overlooked your PR and have not checked where the options were located.

**Comment by cary-rowen** [comment]
Hi @mltony 
I tested this and it worked with my bluetooth headphones, and in the past I needed to use the bluetooth audio add-on.

Thanks

**Comment by beqabeqa473** [comment]
these all should be configurable to not play silence or whitenoise,
because most of the wireless headsets have a possibility to turn them
all upon inactivity time.

On 1/28/24, Rowen ***@***.***> wrote:
> Hi @mltony
> I tested this and it worked with my bluetooth headphones, and in the past I
> needed to use the bluetooth audio add-on.
>
> Thanks
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/16099#issuecomment-1913450600
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by mltony** [comment]
@beqabeqa473

Silence duration and white noise volume are both configurable.

> most of the wireless headsets have a possibility to turn them all upon inactivity time.

Not sure I understand this comment. Most Bluetooth headphones enter standby mode soon after each utterance  which causes speech dropout. As far as I know there is typically no way to configure standby mode on vast majority of Bluetooth headphones.

**Comment by beqabeqa473** [comment]
No, i ment 2.4G wireless headset, they are beeing turned off after
some period of inactivity which is configurable, it is 10 minutes for
me, and i don't this to go away, that is why i said that this options
should be configurable to turn them all together.

On 1/28/24, mltony ***@***.***> wrote:
> @beqabeqa473
>
> Silence duration and white noise volume are both configurable.
>
>> most of the wireless headsets have a possibility to turn them all upon
>> inactivity time.
>
> Not sure I understand this comment. Most Bluetooth headphones enter standby
> mode soon after each utterance  which causes speech dropout. As far as I
> know there is typically no way to configure standby mode on vast majority of
> Bluetooth headphones.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/16099#issuecomment-1913464386
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by seanbudd** [comment]
I believe the white noise parameter should not be settable from the GUI, as its for dev debugging only, but keeping it has a hidden setting in nvda.ini may be appropriate

**Comment by mltony** [comment]
@seanbudd, I just dropped white noise volume from settings panel.

**Comment by cary-rowen** [comment]
If the white noise volume is only used for development and debugging, it should be recorded in an appropriate doc. Otherwise, even developers will not know how to use it because it is too hidden.

**Comment by cary-rowen** [comment]
If I remember correctly, reporting progress also has a hidden interval that can be configured.
This information should be documented, such as in the "Advanced Topics" section of the user guide.

File: source/gui/settingsDialogs.py [hunk 3]
```diff
@@ -3295,6 +3295,23 @@ def __init__(self, parent):
 
 		self.Layout()
 
+		silenceDurationLabelText = _(
```
Comment (CyrilleB79) [inline source/gui/settingsDialogs.py:3298]: There is an "audio" grouping in the advanced settings panel. Could you add this parameter there rather than at the end? And update the documentation the same way?

File: source/gui/settingsDialogs.py [hunk 4]
```diff
@@ -3295,6 +3295,23 @@ def __init__(self, parent):
 
 		self.Layout()
 
+		silenceDurationLabelText = _(
+			# Translators: The label for a setting in advanced panel to change silence duration
+			"Duration in seconds of silence or white noise played to keep audio device open (or 0 to disable)"
```
Comment (CyrilleB79) [inline source/gui/settingsDialogs.py:3300]: 1. I would totally drop "white noise" from the GUI. I'd recommend to add it in the developer's guide with other hidden config parameters (maybe in a separate PR though).
2. It's quite obvious that setting to 0 will add no silence at all. You can add "0 to disable" in the User Guide, but IMO, the GUI should not be cluttered with this detail.

```suggestion
			"Duration in seconds of silence to keep audio device open"
```

File: source/core.py [hunk 5]
```diff
@@ -946,6 +945,7 @@ def _doPostNvdaStartupAction():
 			)
 		except:
 			pass
+	_terminate(nvwave)
```
Comment (CyrilleB79) [inline source/core.py:948]: Why was it moved? Could you update the initial description to give the reason if it is needed?

File: user_docs/en/changes.t2t [hunk 6]
```diff
@@ -8,9 +8,9 @@ What's New in NVDA
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
 - Added support for the BrailleEdgeS2 braille device. (#16033)
+- Playing silence after every utterance in order to keep audio device open (#14386, @jcsteh, @mltony)
```
Comment (CyrilleB79) [inline user_docs/en/changes.t2t:11]: Wouldn't you put something more user oriented?

E.g.:
```suggestion
- NVDA will play silence after every utterance in order to prevent the start of next speech to be clipped with some audio devices such as Bluetooth headphones. (#14386, @jcsteh, @mltony)
```

But this may probably be reworded still better.

File: user_docs/en/userGuide.t2t [hunk 7]
```diff
@@ -2584,6 +2584,15 @@ Disabling WASAPI will disable the following options:
 - [Volume of NVDA sounds #SoundVolume]
 -
 
+==== Silence duration ====[silenceDuration]
+
+This edit box specifies how long NVDA plays silence after every utterance. We play silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
```
Comment (CyrilleB79) [inline user_docs/en/userGuide.t2t:2589]: we -> NVDA

```suggestion
This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
```

**Comment by CyrilleB79** [comment]
And also a question:
In [Jaws What's new](https://support.freedomscientific.com/Downloads/JAWS/JAWSWhatsNew?version=2020) presenting of the similar feature, I can read:
> This option is also not available if the Audio Ducking feature is enabled which lowers the audio volume of programs while JAWS speaks.

Have you checked this here? Is there a side effect?

**Comment by XLTechie** [comment]
I have a question for @mltony or others:

I do not need this at all on my internal speakers. They have their 
own power saving feature, but it doesn't interfere with speech.

However, I do have bluetooth devices with very aggressive power saving, where 
this feature is needed. (and works well, thanks @mltony!)

It is both potentially costly to my laptop battery, and unnecessary, to use this 
feature when using the on-board speakers.

My question therefore is: could or should this be made configurable on a 
per-device basis?
I'm not sure if that would be tricky in the context of the Windows sound mapper, 
but not even most of my bluetooth headphones need this feature. Thus it would be 
nice to deploy it to only those which do.

That said, even if this is thought valid, I wouldn't suggest delaying the PR to 
achieve it. I'm also not sure how the UI for this might look.

**Comment by CyrilleB79** [comment]
I wonder if the effect of keeping the audio stream open on the laptop battery is really significant. @jcsteh have you done or heard from such benchmark in the past?

If yes, the issue raised by @XLTechie is completely valid. Though, implementing a per device configuration should be tracked in a separate issue and a subsequent PR IMO.
In the meantime, it will still be possible to assign a bluetooth profile that you may want to assign manually.

And if it is confirmed that the effect is significant on the laptop battery of some configurations, I'd say that the configuration of this feature should be moved from Advanced panel to Audio panel, unlike what I wrote before.

**Comment by XLTechie** [comment]
@CyrilleB79 I'm not really setup to accurately test the battery of the laptop 
question, but I can probably more effectively figure out a way to test the 
battery of various bluetooth devices which don't need, but are subjected to, 
this feature.

Unfortunately I'm still recovering from some health problems I've had since 
Christmas, so I doubt I can do it in the next several days.

**Comment by mltony** [comment]
@CyrilleB79 
Just checked ducking - appears to work fine with sound split.
@XLTechie
That's a valid argument. I would think that the impact on your laptop battery life should be minimal with default 30 second silence value. But feel free to test that.

File: nvdaHelper/local/wasapi.cpp [hunk 8]
```diff
@@ -624,9 +634,128 @@ HRESULT WasapiPlayer::setChannelVolume(unsigned int channel, float level) {
...
+	UINT32 n = whiteNoiseData.size();
+	const double mean = 0.0;
+	const double stddev = volume * 256;
+	std::default_random_engine generator;
+	std::normal_distribution<double> dist(mean, stddev);
+	for (UINT32 i = 0; i < n; i++) {
+		whiteNoiseData[i] = (INT16)dist(generator);
+	}
+}
+HRESULT SilencePlayer::init() {
```
Comment (seanbudd) [inline nvdaHelper/local/wasapi.cpp:705]:
```suggestion
}

HRESULT SilencePlayer::init() {
```

File: source/config/configSpec.py [hunk 9]
```diff
@@ -54,6 +54,8 @@
 	WASAPI = featureFlag(optionsEnum="BoolFlag", behaviorOfDefault="enabled")
 	soundVolumeFollowsVoice = boolean(default=false)
 	soundVolume = integer(default=100, min=0, max=100)
+	silenceTimeSeconds = integer(default=30,min=0,max=3600)
```
Comment (seanbudd) [inline source/config/configSpec.py:57]:
```suggestion
	silenceTimeSeconds = integer(default=30, min=0, max=3600)
```

File: user_docs/en/changes.t2t [hunk 10]
```diff
@@ -8,9 +8,9 @@ What's New in NVDA
 == New Features ==
 - In Windows 11, NVDA will announce alerts from voice typing and suggested actions including the top suggestion when copying data such as phone numbers to the clipboard (Windows 11 2022 Update and later). (#16009, @josephsl)
 - Added support for the BrailleEdgeS2 braille device. (#16033)
+- NVDA will play silence after every utterance in order to prevent the start of next speech to be clipped with some audio devices such as Bluetooth headphones. (#14386, @jcsteh, @mltony)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:11]:
```suggestion
- NVDA will play silence after every utterance in order to prevent the start of the next speech being clipped with some audio devices such as Bluetooth headphones. (#14386, @jcsteh, @mltony)
```

**Comment by seanbudd** [comment]
What are your thoughts on rephrasing this in the UX to "keep audio device awake" rather than "play silence"?

I think this is a bit more user friendly and outlines in the intended purpose better.
Thoughts @Qchristensen ?

**Comment by Qchristensen** [comment]
> ...rephrasing this in the UX to "keep audio device awake" rather than "play silence"...

I'm in favour of this change.  As you note, "keep audio device awake" is descriptive of WHAT will happen if I enable that option.  "Play silence" is the HOW NVDA will do that, which, while some users may be interested in that, isn't as clear.  (It may be in the future there is some better way of keeping the audio device awake which doesn't involve explicitly sending silent sounds to it, in which case the first option will still be correct, but the second may not be.

File: user_docs/en/userGuide.t2t [hunk 11]
```diff
@@ -2584,6 +2584,15 @@ Disabling WASAPI will disable the following options:
 - [Volume of NVDA sounds #SoundVolume]
 -
 
+==== Silence duration ====[SilenceDuration]
+
+This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
```
[the remark this answers is not in the document]
Comment (CyrilleB79) [reply user_docs/en/userGuide.t2t:2589]: Given the previous suggestion, we should replace "playing audio" by "keeping audio awake".

File: source/config/configSpec.py (hunk 9 above)
Comment (CyrilleB79) [inline source/config/configSpec.py:57]: If you accept to replace "play silence" by "Keep audio device awake" in the User Guide and the GUI, you may also change it here for consistency:

```suggestion
	keepAudioAwakeTimeSeconds = integer(default=30,min=0,max=3600)
```

**Review by CyrilleB79 — COMMENTED** [review]
I fully agree with replacing "Play silence" by "Keep audio device awake" in the GUI and in the User Guide for the reason indicated by @Qchristensen

**Comment by mltony** [comment]
Renamed to "keep audio device awake" in UX as suggested. In user guide I renamed the section as well, but I left explanation that this is achieved by playing silence. Hope that's ok.

**Comment by Adriani90** [comment]
@mltony and @jcsteh thanks very much for this nice work guys.

in the user guide, I think it makes sense to replace the word utterance with "speech sequence". This is easier to understand especially for people who need easier language. @mltony what do you think?
Also I would mention that this does not affect only bluetooth devices positively, but also systems with lower performance, and virtual machines such as virtual desktops in Citrix environments.

**Comment by mltony** [comment]
@Adriani90, updated utterance -> speech sequence.
Re other use cases, what evidence do we have that silence positively affects systems with lower performance? My wife is running NVDA on a modest 8-year old laptop and she doesn't experience any problems with NVDA audio. I guess there must be users who are running NVDA on even older computers. But then I'd think just being low performance wouldn't necessarily imply bad audio quality - NVDA in general doesn't consume  much resources at all. Maybe there is some old technology like a specific brand of motherboards that suffers from this problem? I don't know whether we should go into that rabbithole of investigating which old equipment causes dropout in NVDA speech, but I would prefer to have some evidence (like user testimonial) that silence indeed helps on their old equipment.
Same thing for citrix. I previously ran NVDA in VMWare Player and  VirtualBox and I recall audio quality wasn't that perfect, but it didn't improve from Bluetooth Audio add-on. And the quality problems wer different: every few seconds there was some audible glitch in sound instead of beginnings of each utterance dropping, which would have been a sign of audio device being closed. My guess was that there was high latency between host and guest that caused occasional dropout of some small chunk of audio buffer. And in my understanding this own't be fixed by this PR.

File: user_docs/en/userGuide.t2t [hunk 12]
```diff
@@ -2605,14 +2605,17 @@ Disabling WASAPI will disable the following options:
 - [Volume of NVDA sounds #SoundVolume]
 -
 
-==== Silence duration ====[SilenceDuration]
+==== Duration of keeping audio device awake ====[KeepAudioAwakeDuration]
 
-This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
+This edit box specifies how long NVDA keeps audio device awake after every speech sequence.
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:2610]:
```suggestion
This edit box specifies how long NVDA keeps the audio device awake after every speech sequence.
```

File: user_docs/en/userGuide.t2t [hunk 13]
```diff
@@ -2605,14 +2605,17 @@ Disabling WASAPI will disable the following options:
 - [Volume of NVDA sounds #SoundVolume]
 -
 
-==== Silence duration ====[SilenceDuration]
+==== Duration of keeping audio device awake ====[KeepAudioAwakeDuration]
 
-This edit box specifies how long NVDA plays silence after every utterance. NVDA plays silence to keep audio device open and avoid certain glitches like dropped parts of utterances, that happen due to audio devices (especially Bluetooth and wireless devices) entering stand by mode.
+This edit box specifies how long NVDA keeps audio device awake after every speech sequence.
+NVDA does that by playing silence to keep audio device open and avoid certain glitches like dropped parts of speech sequences.
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:2611]:
```suggestion
NVDA does this by playing silence to keep the audio device open and avoid certain glitches like dropped parts of speech sequences.
```

File: user_docs/en/userGuide.t2t [hunk 14]
```diff
@@ -2605,14 +2605,17 @@ Disabling WASAPI will disable the following options:
...
+This can happen due to audio devices (especially Bluetooth and wireless devices) entering stand-by mode.
 
-* Too low value would mean  audio may be cut-off more often.
-* Too high value would mean the battery of sound output device discharges faster.
+Lower values may mean audio would be cut-off more often, as a device may enter stand-by mode, causing the next speech to be clipped.
+Too high a value may mean the battery of the sound output device discharges faster, as it stays active for longer while no sound is being sent.
 
-You can specify silence duration to be 0 in order to disable this feature. This slider is only availabel when wasapi mode is enabled.
+You can specify silence duration to be 0 in order to disable this feature.
+This slider is only available when WASAPI mode is enabled.
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:2618]:
```suggestion
This slider is only available when [WASAPI mode #WASAPI] is enabled.
```

**Review by seanbudd — APPROVED** [review]

**Comment by Adriani90** [comment]
At least in Citrix I am using NVDA at work and without the bluetooth addon the speech is cutting and crackling alot. I am not using any wireless headset, but a wired headfone. The bluetooth addon definitely has a positive effect on this. Without it the citrix virtual desktop is literarly unusable for me.  Von meinem iPhone gesendetAm 07.02.2024 um 23:38 schrieb mltony ***@***.***>:﻿
@Adriani90, updated utterance -> speech sequence.
Re other use cases, what evidence do we have that silence positively affects systems with lower performance? My wife is running NVDA on a modest 8-year old laptop and she doesn't experience any problems with NVDA audio. I guess there must be users who are running NVDA on even older computers. But then I'd think just being low performance wouldn't necessarily imply bad audio quality - NVDA in general doesn't consume  much resources at all. Maybe there is some old technology like a specific brand of motherboards that suffers from this problem? I don't know whether we should go into that rabbithole of investigating which old equipment causes dropout in NVDA speech, but I would prefer to have some evidence (like user testimonial) that silence indeed helps on their old equipment.
Same thing for citrix. I previously ran NVDA in VMWare Player and  VirtualBox and I recall audio quality wasn't that perfect, but it didn't improve from Bluetooth Audio add-on. And the quality problems wer different: every few seconds there was some audible glitch in sound instead of beginnings of each utterance dropping, which would have been a sign of audio device being closed. My guess was that there was high latency between host and guest that caused occasional dropout of some small chunk of audio buffer. And in my understanding this own't be fixed by this PR.

**Comment by mltony** [comment]
@Adriani90 , fair enough, I added a line about that in the user guide. Now I also vaguely remember someone on NVDA mailing list was saying their laptop's built-in audio was suffering from audio cutoffs and Bluetooth add-on indeed helped, so I mentioned that as well.

**Review by Qchristensen — APPROVED** [review]
All reads well, good work.

**Comment by jcsteh** [comment]
@mltony Thanks very much for driving this to completion. :)

**Comment by XLTechie** [comment]
I'm curious why it is thought that "speech sequence" makes sense to end users? 
Wouldn't something like "each spoken phrase", or describing it as "after speech 
stops" be more comprehensible?

Even the original, "utterance", comes across as a more user friendly word, 
although perhaps some other languages have issues with it as @Adriani90 implied.

I would rather see us describe it with as few technical terms as possible. 
"Speech sequence" strikes me as a term of art, which is not well defined 
outside our developer community.
